mlflow/torchscript.py

```diff
@@ -0,0 +1,313 @@
+                      contained in :func:`get_default_conda_env()`. If ``None``, the default
+                      :func:`get_default_conda_env()` environment is added to the model. The
+                      following is an *example* dictionary representation of a Conda environment::
+
+                        {
+                            'name': 'mlflow-env',
+                            'channels': ['defaults', 'pytorch'],
+                            'dependencies': [
+                                'python=3.7.0',
+                                'pytorch=1.5.0',
+                                'torchvision=0.2.1'
+                            ]
+                        }
+
+    :param mlflow_model: :py:mod:`mlflow.models.Model` this flavor is being added to.
+
+    :param signature: (Experimental) :py:class:`ModelSignature <mlflow.models.ModelSignature>`
+                      describes model input and output :py:class:`Schema <mlflow.types.Schema>`.
+                      The model signature can be :py:func:`inferred <mlflow.models.infer_signature>`
+                      from datasets with valid model input (e.g. the training dataset with target
+                      column omitted) and valid model output (e.g. model predictions generated on
+                      the training dataset), for example:
+
+                      .. code-block:: python
+
+                        from mlflow.models.signature import infer_signature
+                        train = df.drop_column("target_label")
+                        predictions = ... # compute model predictions
+                        signature = infer_signature(train, predictions)
+    :param input_example: (Experimental) Input example provides one or several instances of valid
+                          model input. The example can be used as a hint of what data to feed the
+                          model. The given example will be converted to a Pandas DataFrame and then
+                          serialized to json using the Pandas split-oriented format. Bytes are
+                          base64-encoded.
+
+    :param kwargs: kwargs to pass to ``torch.jit.save`` method.
+
+    .. code-block:: python
+        :caption: Example
+
+        import torch
+        import mlflow
+        import mlflow.torchscript
+        # Create model and set values
+        model = Model()
+        # train our model
+        for epoch in range(500):
+            y_pred = model(x_data)
+            ...
+        # Save the model
+        with mlflow.start_run() as run:
+            mlflow.log_param("epochs", 500)
+            scripted_model = torch.jit.script(model)
+            mlflow.torchscript.save_model(scripted_model, pytorch_model_path)
+    """
+    if not hasattr(model, 'state_dict') or not hasattr(model, 'save'):
+        # If it walks like a duck and it quacks like a duck, then it must be a duck
```

Comment:
Could we replace this with:

```
if not isinstance(model, torch.jit.ScriptModule) and not isinstance(model, torch.jit.ScriptFunction):
            raise TypeError("Argument 'model' should be a TorchScript model. Use torch.jit.script() to convert your model into a TorchScript model before passing it to this function.")
```

It looks like `torch.jit.script`'s documented behavior is to return either a `torch.jit.ScriptModule` or `torch.jit.ScriptFunction`: https://pytorch.org/docs/stable/jit.html#torch.jit.script

Reply:
Done